### PR TITLE
Pane renaming saves the name of the panes into their state

### PR DESCRIPTION
**Pull request by dead-tech**
~~Now changes made to pane names are saved to local storage, and should be working as expected.~~

Save name of panes into their state.

Note  (See #3323):
 - Renaming for editor is still disabled at the moment
 - `Alert` still isn't a singletone.

### CODE REVIEW

File: static/pane-renaming.ts [hunk 1]
```diff
@@ -25,32 +25,55 @@
...
-        // pane.container.on('tab', addRenameButton);
+export class PaneRenaming extends EventEmitter.EventEmitter{
+    private pane: any;
+    private alertSystem: any;
+    private state: any;
+
+    constructor(pane: any, state: any) {
+        super();
+        this.pane = pane;
+        this.alertSystem = new Alert();
```
Comment (RubenRBS) [inline static/pane-renaming.ts:40]: What you can do here about the alerty situation is to use the alert coming from the parent Pane (In most panes, it's the `this.alertSystem variable`) So something like this

```suggestion
    constructor(pane: any, state: any, alertSystem?: Alert) {
        super();
        this.pane = pane;
        if (alertSystem === undefined) {
            this.alertSystem = new Alert();
        } else {
            this.alertSystem = alertSystem;
        }
```

**Review by RubenRBS — CHANGES_REQUESTED** [review]
Looking good. Thanks for taking the time to fix the issues found with the initial version :)

Even though it's currently disabled, can you update `Editor.prototype.updateTitle` to check for `'CMakeLists.txt'` always in the result from `getPaneName` and not the custom name even if it has one please?

File: static/pane-renaming.ts [hunk 2]
```diff
@@ -25,32 +25,55 @@
 import $ from 'jquery';
```
[the remark this answers is not in the document]
Comment (dead-tech) [reply static/pane-renaming.ts:25]: Good to know 😄

File: static/panes/ast-view.js [hunk 3]
```diff
@@ -90,6 +91,7 @@ Ast.prototype.initCallbacks = function () {
     }, this));
 
     this.fontScale.on('change', _.bind(this.updateState, this));
+    this.paneRenaming.on('renamePane', this.updateState.bind(this));
```
[the remark this answers is not in the document]
Comment (dead-tech) [reply static/panes/ast-view.js:94]: So I guess it's fine if I leave them as they are.

File: static/panes/cfg-view.interfaces.ts [hunk 4]
```diff
@@ -32,4 +32,5 @@ export interface CfgState {
         physics?: boolean;
         navigation?: boolean;
     }
+    componentName: string;
```
[the remark this answers is not in the document]
Comment (dead-tech) [reply static/panes/cfg-view.interfaces.ts:35]: My bad, it was something that had to do with the local storage version of the feature. Removed it!

File: static/panes/pp-view.ts [hunk 5]
```diff
@@ -125,7 +125,8 @@ export class PP extends Pane<monaco.editor.IStandaloneCodeEditor, PPViewState> {
     }
 
     override updateTitle() {
-        this.container.setTitle(this.getPaneName());
+        const name = this.paneName ?? this.getPaneName();
+        this.container.setTitle(_.escape(name));
```
[the remark this answers is not in the document]
Comment (dead-tech) [reply static/panes/pp-view.ts:129]: Yeah, they're identical, removed the overridden one.

**Comment by dead-tech** [comment]
> Looking good. Thanks for taking the time to fix the issues found with the initial version :)
> 
> Even though it's currently disabled, can you update `Editor.prototype.updateTitle` to check for `'CMakeLists.txt'` always in the result from `getPaneName` and not the custom name even if it has one please?

@RubenRBS Is this ([ba7e0b7](https://github.com/compiler-explorer/compiler-explorer/pull/3364/commits/ba7e0b7adec56f284c2cc00f8a3bd0102abb8624)) what you meant?

**Comment by RubenRBS** [comment]
Yes! Thanks

**Comment by dead-tech** [comment]
@mattgodbolt on #3360 said this:

> I think it's considered bad form to dynamically add members in this way: this is also a problem when `this.paneName = "";` is equivalent to `this.paneName` was never defined at all (which may be intentional, but again is subtle). I'd prefer I think explicitly initialising it to `null` (or `""` if that makes sense) in the constructor to say "this is at least expected to be here".

So shall I go ahead and define `paneName` and set it to `null` in the constructor of all the panes that expect this property to exist, and not dynamically add this?

Btw as #3360 is waiting for this pr to get merged, I would say, apart from the change above, this might be ready

**Comment by RubenRBS** [comment]
Hi! Sorry for the review delay. I plan to test this later this evening :)

File: static/panes/cfg-view.ts [hunk 6]
```diff
@@ -30,6 +30,7 @@ import TomSelect from 'tom-select';
 import { Container } from 'golden-layout';
 import { CfgState } from './cfg-view.interfaces';
 import { PaneRenaming } from '../pane-renaming';
+const Alert = require('../alert').Alert;
```
Comment (RubenRBS) [inline static/panes/cfg-view.ts:33]:
```suggestion
import { Alert } from '../alert';
```

File: static/panes/pane.interfaces.ts [hunk 7]
```diff
@@ -42,5 +42,6 @@ export interface BasePaneState {
     id: number;
     compilerName: string;
     editorid: number;
+    componentName: string;
```
Comment (RubenRBS) [inline static/panes/pane.interfaces.ts:45]: Is this still needed?

**Review by RubenRBS — APPROVED** [review]

File: static/panes/pane.interfaces.ts (hunk 7 above)
Comment (dead-tech) [reply static/panes/pane.interfaces.ts:45]: Nope, my bad

**Comment by RubenRBS** [comment]
Sorry that this took so long to get reviewed again, life got busy.
Thank you so much for taking the time to implement this feature and again, sorry about all the delays along the way :)

**Comment by mattgodbolt** [comment]
This is now live! Yay!

**Pull request merged**